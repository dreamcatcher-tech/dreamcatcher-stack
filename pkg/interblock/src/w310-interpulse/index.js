export * from './src/Interpulse'
export * from './src/Crisp'
export * from './src/Syncer'
