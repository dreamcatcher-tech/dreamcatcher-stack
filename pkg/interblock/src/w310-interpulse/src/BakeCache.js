import assert from 'assert-fast'
import posix from 'path-browserify'
import Debug from 'debug'
import { Address, Pulse, PulseLink } from '../../w008-ipld'
import { Map as IMap } from 'immutable'
import { pushable } from 'it-pushable'

const debug = Debug('interblock:api:BakeCache')

/**
 * Format is: chainId -> {
 *    pulseId, // latest known pulseId, which may be ahead of the pulse
 *    pulse, // the last fully walked pulse, or the first unwalked pulse
 *    channels<Map<channelId, channel>>, // the last fully walked channels, or the first progressively walked channels
 *    done // true if the pulse has been fully walked, so it can be used as a prior
 * }
 *
 * pulse and channels are a matched pair
 */
export class BakeCache {
  #yieldStream
  #chains = new Map() // cids -> { pulseId, pulse, channels, done }
  #covenants = new Map() // paths -> pulse
  static create(yieldStream) {
    assert.strictEqual(typeof yieldStream.push, 'function')
    const cache = new BakeCache()
    cache.#yieldStream = yieldStream
    return cache
  }
  static createCI() {
    const yieldStream = pushable({ objectMode: true })
    return BakeCache.create(yieldStream)
  }
  setCovenant(path, pulse) {
    assert(posix.isAbsolute(path), `path must be absolute: ${path}`)
    assert(pulse instanceof Pulse)
    assert(!this.#covenants.has(path))
    this.#covenants.set(path, pulse)
    this.#yieldStream.push({ type: 'COVENANT', path, pulse })
  }
  preBake(address, pulseId) {
    assert(address instanceof Address)
    assert(pulseId instanceof PulseLink)
    const key = address.getChainId()
    if (!this.#chains.has(key)) {
      this.#chains.set(key, {})
    }
    const tracker = this.#chains.get(key)
    if (!pulseId.equals(tracker.pulseId)) {
      this.#chains.set(key, { ...tracker, pulseId })
      this.#yieldStream.push({ type: 'PRE_BAKE', address, pulseId })
    }
  }
  isBaked(address, pulseId) {
    assert(address instanceof Address)
    assert(pulseId instanceof PulseLink)
    const key = address.getChainId()
    assert(this.#chains.has(key), `not initialized: ${address}`)
    const tracker = this.#chains.get(key)
    assert(pulseId.equals(tracker.pulseId))
    if (tracker.pulse) {
      if (!tracker.pulse.getPulseLink().equals(pulseId)) {
        return false
      }
      return this.isWalked(address)
    }
    return false
  }
  hasPulse(address) {
    assert(address instanceof Address)
    const key = address.getChainId()
    if (!this.#chains.has(key)) {
      return false
    }
    const { pulse } = this.#chains.get(key)
    return !!pulse
  }
  getPulse(address) {
    assert(address instanceof Address)
    const key = address.getChainId()
    const { pulse } = this.#chains.get(key)
    assert(pulse instanceof Pulse, `no pulse for ${address}`)
    return pulse
  }
  setVirginPulse(address, pulse) {
    assert(address instanceof Address)
    assert(pulse instanceof Pulse)
    const key = address.getChainId()
    assert(this.isVirgin(address), `already baked: ${address}`)
    const tracker = this.#chains.get(key)
    const { pulseId } = tracker
    assert(pulseId instanceof PulseLink)
    assert(pulseId.equals(pulse.getPulseLink()))
    this.#chains.set(key, { ...tracker, pulse })
    this.#yieldStream.push({ type: 'VIRGIN_PULSE', address, pulse })
  }
  isVirgin(address) {
    assert(address instanceof Address)
    const key = address.getChainId()
    assert(this.#chains.has(key), `not initialized: ${address}`)
    const { pulse, done, channels } = this.#chains.get(key)
    return !pulse && !done && !channels
  }
  isWalked(address) {
    assert(address instanceof Address)
    const key = address.getChainId()
    assert(this.#chains.has(key), `not initialized: ${address}`)
    const { done, pulse, channels } = this.#chains.get(key)
    if (done) {
      assert(pulse instanceof Pulse)
      assert(IMap.isMap(channels))
    }
    return !!done
  }
  hasChannels(address) {
    assert(address instanceof Address)
    const key = address.getChainId()
    assert(this.#chains.has(key), `not initialized: ${address}`)
    const { channels } = this.#chains.get(key)
    return !!channels
  }
  getChannels(address) {
    assert(address instanceof Address)
    const key = address.getChainId()
    assert(this.#chains.has(key), `not initialized: ${address}`)
    const { channels } = this.#chains.get(key)
    assert(IMap.isMap(channels))
    return channels
  }
  updateChannels(address, channels) {
    assert(address instanceof Address)
    assert(IMap.isMap(channels))
    const key = address.getChainId()
    const { done, ...rest } = this.#chains.get(key)
    assert(!done, `already done: ${address}`)
    this.#chains.set(key, { ...rest, channels })
    this.#yieldStream.push({ type: 'UPDATE_CHANNELS', address })
  }
  finalize(address, pulse, channels) {
    assert(address instanceof Address)
    assert(pulse instanceof Pulse)
    assert(IMap.isMap(channels))
    const key = address.getChainId()
    const { pulseId } = this.#chains.get(key)
    assert(pulseId.equals(pulse.getPulseLink()))
    this.#chains.set(key, { pulseId, pulse, channels, done: true })
    this.#yieldStream.push({ type: 'FINALIZE', address, pulse })
  }
  hasCovenant(path) {
    assert(posix.isAbsolute(path), `path must be absolute: ${path}`)
    return this.#covenants.has(path)
  }
  getCovenant(path) {
    assert(posix.isAbsolute(path), `path must be absolute: ${path}`)
    assert(this.#covenants.has(path), `no covenant for path: ${path}`)
    return this.#covenants.get(path)
  }
}
