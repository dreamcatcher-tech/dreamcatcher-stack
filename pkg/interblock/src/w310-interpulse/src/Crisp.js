import assert from 'assert-fast'
import Debug from 'debug'
import { Pulse } from '../../w008-ipld'
import Immutable from 'immutable'
const debug = Debug('interblock:api:Crisp')

export class Crisp {
  #parent // the parent Crisp
  // #appRoot // Pulse that is the root of the app
  #pulse // the Pulse that we wrap
  static createRoot(rootPulse) {
    assert(rootPulse instanceof Pulse)
    const result = new Crisp()
    result.#pulse = rootPulse
    return result
  }
  static createChild(pulse, parent) {
    assert(pulse instanceof Pulse)
    assert(parent instanceof Crisp)
    const result = new Crisp()
    result.#pulse = pulse
    result.#parent = parent
    return result
  }
  get root() {
    // walk the parents up until we have root
    if (!this.#parent) {
      return this
    }
    return this.#parent.root
  }
  get isActionsLoaded() {
    // returns true if the actions are fully loaded
    // this call signals the reconciler that loading is required
    return false
  }
  get parent() {
    return this.#parent
  }
  get state() {
    const state = this.#pulse.getState().toJS()
    return state
  }
  getChild(path) {
    assert.strictEqual(typeof path, 'string')
    if (!this.hasChild(path)) {
      throw new Error(`child not found: ${path}`)
    }
  }
  hasChild(path) {
    assert.strictEqual(typeof path, 'string')
    const network = this.#pulse.getNetwork()

    //if we have this child in whatever is loaded, return true
    // if we are not loaded, return false
    // if we are loaded, but do not have this path, return false

    // this will signal the reconciler to load this path

    // first load the channel into our local cache to get the tip hash
    // then load the pulse into the pulsecache
    // then return true when this is asked for again
  }
  #mapView
  get #map() {
    if (!this.#mapView) {
      const network = this.#pulse.getNetwork()
      this.#mapView = network.channels.list.bakedMap ?? Immutable.Map()
    }
    return this.#mapView
  }
  *[Symbol.iterator]() {
    for (const [key, value] of this.#map.entries()) {
      yield [key, value]
    }
    yield ''
  }
}
