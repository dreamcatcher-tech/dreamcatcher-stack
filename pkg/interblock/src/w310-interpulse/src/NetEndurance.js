import { CID } from 'multiformats/cid'
import { PulseLink } from '../../w008-ipld/index.mjs'
import assert from 'assert-fast'
import { Endurance } from '../../w210-engine'
import { PulseNet } from '../../w305-libp2p'
import Debug from 'debug'
const debug = Debug('interblock:interpulse:NetEndurance')

export class NetEndurance extends Endurance {
  static async create(pulseNet) {
    assert(pulseNet instanceof PulseNet)
    const instance = new NetEndurance()
    instance.#net = pulseNet
    await instance.#start()
    return instance
  }
  #net
  #writePromise = Promise.resolve()
  #writeResolve
  #writeCount = 0
  #writeStart() {
    if (!this.#writeCount) {
      this.#writePromise = new Promise((resolve) => {
        this.#writeResolve = resolve
      })
    }
    this.#writeCount++
  }
  #writeStop() {
    this.#writeCount--
    if (!this.#writeCount) {
      this.#writeResolve()
      this.#writeResolve = undefined
      this.#writePromise = Promise.resolve()
    }
  }
  async #start() {
    // TODO retransmit the tx's of latest pulse
    const { repo } = this.#net
    assert(!repo.closed)
    const config = await repo.config.getAll()
    if (config.latest) {
      const pulselink = PulseLink.parse(config.latest)
      const latest = await this.recover(pulselink)
      await super.endure(latest) // acts like the bootstrap pulse
    }
  }
  async endure(latest) {
    let isBootstrapPulse = !this.selfAddress
    await super.endure(latest)
    this.#writeStart()
    const result = this.#net.endure(latest)
    debug(`start ipfs put`, latest.getPulseLink())
    result.then(() => {
      debug(`finish net endure`, latest.getPulseLink())
      this.#writeStop()
    })
    if (isBootstrapPulse || this.selfAddress.equals(latest.getAddress())) {
      const pulselink = latest.getPulseLink().cid.toString()
      this.#net.repo.config.set('latest', pulselink)
    }
  }
  async recover(pulselink) {
    const result = await super.recover(pulselink)
    if (result) {
      return result
    }
    return await this.#net.getPulse(pulselink)
  }
  getResolver(treetop) {
    assert(CID.asCID(treetop))
    const netResolver = this.#net.getResolver(treetop)
    const resolver = super.getResolver(treetop)
    // TODO WARNING permissions must be honoured
    // TODO use treetop to only fetch things below this CID
    return async (cid) => {
      const result = await resolver(cid)
      if (result) {
        return result
      }
      return await netResolver(cid)
    }
  }
  async stop() {
    super.stop()
    await this.#writePromise
  }
  async scrub(pulse, { history } = {}) {
    // walk the pulse, its interpulses, and optionally its history and binaries
  }
  async fade(pulse) {
    // remove the pulse from local storage whenever next convenience arises
  }
}
