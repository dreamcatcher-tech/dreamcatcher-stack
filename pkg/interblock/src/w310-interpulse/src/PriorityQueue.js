import assert from 'assert-fast'
import { pushable } from 'it-pushable'

export class PriorityQueue {
  #buffer = new Set()
  #trigger = pushable({ objectMode: true })
  static create() {
    return new PriorityQueue()
  }
  push(id) {
    assert.strictEqual(typeof id, 'string')
    if (!this.#buffer.has(id)) {
      this.#trigger.push({})
    }
    this.#buffer.delete(id)
    this.#buffer.add(id)
  }
  end(error) {
    this.#trigger.end(error)
  }
  async *[Symbol.asyncIterator]() {
    for await (const trigger of this.#trigger) {
      const iterator = this.#buffer[Symbol.iterator]()
      const id = iterator.next().value
      iterator.return()
      this.#buffer.delete(id)
      yield id
    }
  }
}
