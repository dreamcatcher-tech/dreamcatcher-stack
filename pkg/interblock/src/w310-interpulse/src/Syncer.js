import parallel from 'it-parallel'
import { pipe } from 'it-pipe'
import assert from 'assert-fast'
import posix from 'path-browserify'
import Debug from 'debug'
import {
  Dmz,
  Hamt,
  IpldInterface,
  Pulse,
  PulseLink,
  HistoricalPulseLink,
} from '../../w008-ipld'
import { Crisp } from '..'
import Immutable from 'immutable'
import { pushable } from 'it-pushable'

const debug = Debug('interblock:api:Syncer')
const TYPE = 'crispDeepPulse'
const BAKE_TEAR = 'BAKE_TEAR'

export class Syncer {
  #concurrency = 20
  #pulseResolver
  #covenantResolver
  #actions
  #chroot
  #pulse
  #subscribers = new Set()
  #abort = new AbortController()
  #crisp = Crisp.createLoading()
  #bakeQueue = pushable({ objectMode: true })
  #yieldQueue = pushable({ objectMode: true })
  #totallyBaked = Promise.resolve()

  static create(pulseResolver, covenantResolver, actions, chroot = '/') {
    assert.strictEqual(typeof pulseResolver, 'function')
    assert.strictEqual(typeof covenantResolver, 'function')
    assert.strictEqual(typeof actions, 'object')
    assert.strictEqual(typeof actions.dispatch, 'function')
    assert(posix.isAbsolute(chroot), `chroot must be absolute path: ${chroot}`)

    const syncer = new Syncer()
    syncer.#pulseResolver = pulseResolver
    syncer.#covenantResolver = covenantResolver
    syncer.#actions = actions
    syncer.#chroot = chroot
    syncer.#startYieldQueue()
    return syncer
  }
  async update(pulse) {
    assert(pulse instanceof Pulse)
    assert(!pulse.cid.equals(this.#pulse?.cid))
    this.#tearBake()

    // TODO assert lineage matches
    // TODO permit skips in lineage
    const prior = this.#pulse
    // TODO split the bake cache so pulse does not have to be reference equal

    // priors should walk back to the last one we had data on, not tears
    try {
      this.#pulse = await this.#resolve(pulse.getPulseLink())
      this.#totallyBaked = this.#startBakeQueue()
      this.#pulseBake(pulse.getPulseLink(), prior?.getPulseLink())
      await this.#totallyBaked
    } catch (error) {
      if (error.message === BAKE_TEAR) {
        debug('bake queue torn')
      } else if (error.message === 'Endurance is stopped') {
        debug('endurance stopped')
      } else {
        throw error
      }
    }
  }
  get awaitDeepLoad() {
    return this.#totallyBaked
  }
  #tearBake() {
    debug('tearing bake for %s', this.#pulse)
    this.#bakeQueue?.throw(new Error(BAKE_TEAR))
    this.#bakeQueue = undefined
    this.#abort?.abort('tear bake')
  }
  /**
   * Will mutate instance and all children of instance
   * by expanding any PulseLinks to Pulses, and any Hamts to Maps.
   * @param {IpldInterface | [IpldInterface]} instance
   */
  async #pulseBake(pulse, prior) {
    assert(pulse instanceof PulseLink)
    assert(!prior || prior instanceof PulseLink)
    // bakes pulses one at a time, so we load breadth first
    this.#bakeQueue.push({ pulse, prior })
  }

  async #startBakeQueue() {
    assert(!this.#bakeQueue)
    const queue = pushable({ objectMode: true })
    const abort = new AbortController()
    this.#bakeQueue = queue
    this.#abort = abort
    return new Promise((resolve, reject) => {
      const drain = async () => {
        try {
          for await (const { pulse, prior } of queue) {
            // TODO allow parallel walks
            assert(pulse instanceof PulseLink)
            assert(!pulse.bakedPulse)
            assert(!prior || prior instanceof PulseLink)
            const fullPulse = await this.#resolve(pulse, abort)
            pulse.bake(fullPulse)
            await this.#yield()
            await this.#bake(pulse.bakedPulse, prior?.bakedPulse, abort)

            if (this.#bakeQueue === queue && queue.readableLength === 0) {
              // maybe should only yield once a new pulse has been baked ?
              // not *that* useful to yield partial hamts
              const isDeepLoaded = true
              await this.#yield(isDeepLoaded)
              queue.end()
            }
          }
        } catch (error) {
          if (error.message !== BAKE_TEAR) {
            if (error.message !== 'Endurance is stopped') {
              return reject(error)
            }
          }
        }
        resolve()
      }
      drain()
    })
  }
  async #resolve(pulseLink, abort) {
    assert(pulseLink instanceof PulseLink)
    const pulse = await this.#pulseResolver(pulseLink, TYPE, abort)
    return pulse
  }
  async #bake(instance, prior, abort) {
    assert(abort instanceof AbortController)
    if (abort.signal.aborted) {
      return
    }
    if (Array.isArray(instance)) {
      assert(!prior || Array.isArray(prior))
      return await Promise.all(
        instance.map((v, i) => this.#bake(v, prior?.[i], abort))
      )
    }
    if (!(instance instanceof IpldInterface)) {
      return
    }
    assert(!prior || prior instanceof IpldInterface)
    if (instance instanceof HistoricalPulseLink) {
      return
    }
    if (instance instanceof PulseLink) {
      if (instance.bakedPulse) {
        return
      }
      this.#pulseBake(instance, prior)
      return
    }
    if (instance instanceof Hamt) {
      await this.#updateHamt(instance, prior, abort)
      await this.#yield()
      return
    }
    if (instance instanceof Dmz) {
      await this.#updateCovenant(instance, prior)
      await this.#yield()
    }
    const { classMap = {}, defaultClass } = instance.constructor
    assert(!(Object.keys(classMap).length && defaultClass))

    if (defaultClass) {
      const values = []
      const priorValues = []
      for (const [key, value] of Object.entries(instance)) {
        values.push(value)
        priorValues.push(prior?.[key])
      }
      await this.#bake(values, priorValues, abort)
    } else {
      await Promise.all(
        Object.keys(classMap).map(async (key) => {
          const value = instance[key]
          await this.#bake(value, prior?.[key], abort)
        })
      )
    }
  }
  async #updateCovenant(dmz, prior) {
    // TODO allow slow resolution to not block the rest of the bake
    // TODO is there a point to bake the covenant pulse too ?
    assert(dmz instanceof Dmz)
    assert(!prior || prior instanceof Dmz)
    if (dmz.bakedCovenant) {
      return
    }
    const path = dmz.getCovenantPath()
    const priorPath = prior?.getCovenantPath()

    if (path === priorPath) {
      dmz.bake(prior.bakedCovenant)
      return
    }
    const covenantPulse = await this.#covenantResolver(path)
    dmz.bake(covenantPulse)
  }
  async #updateHamt(hamt, prior, abort) {
    assert(hamt instanceof Hamt)
    assert(!prior || prior instanceof Hamt)
    assert(abort instanceof AbortController)
    if (abort.signal.aborted) {
      return
    }
    if (hamt.isBakeSkippable) {
      return
    }
    let map = hamt.bakedMap ?? prior?.bakedMap ?? Immutable.Map()
    hamt.bake(map)
    const start = Date.now()
    const diff = await hamt.compare(prior) // this definitely locks the event loop
    if (Date.now() - start > 1000) {
      console.log('slow compare', Date.now() - start)
    }
    const { added, deleted, modified } = diff
    assert(map instanceof Immutable.Map)
    for (const key of deleted) {
      map = map.delete(key)
      hamt.bake(map)
      await this.#yield()
    }
    const set = async (key, value) => {
      map = map.set(key, value)
      hamt.bake(map)
      await this.#yield()
    }
    // TODO queue these using emitter hamt walker
    const mods = [...modified].map(async (key) => {
      const value = await hamt.get(key)
      await set(key, value)
      let priorValue
      if (prior) {
        if (prior.bakedMap.has(key)) {
          priorValue = prior.bakedMap.get(key)
        } else {
          priorValue = await prior.get(key)
        }
      }
      await this.#bake(value, priorValue, abort)
    })
    const adds = [...added].map(async (key) => {
      const value = await hamt.get(key)
      await set(key, value)
      await this.#bake(value, undefined, abort)
    })
    await Promise.all([...mods, ...adds])
  }
  async #yield(isDeepLoaded = false) {
    const pulse = this.#pulse
    const actions = this.#actions
    const chroot = this.#chroot
    const args = { pulse, actions, chroot, isDeepLoaded }
    this.#yieldQueue.push(args)
  }
  async #startYieldQueue() {
    for await (const args of this.#yieldQueue) {
      this.#yieldRaw(args)
    }
  }
  #yieldRaw({ pulse, actions, chroot, isDeepLoaded }) {
    // TODO include the prievous crisp in the new crisp
    // TODO only yield if something useful occured, like new child or state
    const crisp = Crisp.createRoot(pulse, actions, chroot, isDeepLoaded)
    this.#crisp = crisp
    for (const subscriber of this.#subscribers) {
      subscriber.push(crisp)
    }
  }
  subscribe() {
    const subscriber = pushable({
      objectMode: true,
      onEnd: () => this.#subscribers.delete(subscriber),
    })
    this.#subscribers.add(subscriber)
    if (this.#crisp) {
      subscriber.push(this.#crisp)
    }
    return subscriber
  }
  set concurrency(value) {
    assert(Number.isInteger(value))
    assert(value > 0, 'concurrency must be greater than 0')
    this.#concurrency = value
  }
}

class BakeCache {
  #pulses = new Map() // cids -> pulse
  #covenants = new Map() // paths -> pulse
  static create() {
    return new BakeCache()
  }
  getCovenant(path) {
    assert(posix.isAbsolute(path), `path must be absolute: ${path}`)
    return this.#covenants.get(path)
  }
  getPulse(pulseLink) {
    assert(pulseLink instanceof PulseLink)
    return this.#pulses.get(pulseLink.cid.toString())
  }
}
