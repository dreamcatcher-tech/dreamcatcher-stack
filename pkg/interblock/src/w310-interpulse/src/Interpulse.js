import { pipe } from 'it-pipe'
import { shell } from '../../w212-system-covenants'
import * as apps from '../../w301-user-apps'
import { pushable } from 'it-pushable'
import { Engine, Endurance } from '../../w210-engine'
import { schemaToFunctions } from '../../w002-api'
import assert from 'assert-fast'
import Debug from 'debug'
import posix from 'path-browserify'
import { PulseNet } from '../../w305-libp2p'
import { NetEndurance } from './NetEndurance'
import { Crypto } from '../../w210-engine/src/Crypto'
import { PulseLink } from '../../w008-ipld/index.mjs'
import { isBrowser, isNode } from 'wherearewe'

const debug = Debug('interpulse')

/**
 * The top level ORM object.
 * Assembles an Engine with all the services it needs to operate.
 *    Where networking is started.
 *    Where multithreading is controlled.
 * Wraps engine with useful functions for devs.
 * Loads the shell to be loaded at the root block.
 * Works with paths, whereas engine works with addresses.
 * Manages subscriptions to remote chains.
 */

export class Interpulse {
  net
  #repo
  #engine
  #endurance
  #crypto
  #subscribers = new Set()

  static async createCI(options = {}) {
    options = { ...options, CI: true }
    return this.create(options)
  }
  static async create(options = {}) {
    let overloads = options.overloads ? { ...options.overloads } : {}
    overloads.root = shell
    Object.assign(overloads, apps)
    // if announcer is inside net, how to trigger on interpulse received ?
    // could connect via Interpulse ?
    let { net, crypto, endurance = Endurance.create() } = options
    const { repo, CI } = options
    if (repo) {
      // no repo => no net - storage and network are one 🙏
      const { tcpHost, tcpPort } = options
      net = await PulseNet.create(repo, CI, tcpHost, tcpPort)
      crypto = Crypto.create(net.keypair)
      endurance = await NetEndurance.create(net)
    }
    const opts = { ...options, overloads, crypto, endurance }
    const engine = await Engine.create(opts)

    const instance = new Interpulse(engine)
    instance.#endurance = endurance
    if (repo) {
      instance.net = net
      instance.#repo = repo
      instance.#crypto = crypto
      instance.#watchMtab()
    }
    return instance
  }
  constructor(engine) {
    assert(engine instanceof Engine)
    const actions = mapShell(engine)
    Object.assign(this, actions)
    this.#engine = engine
  }
  async pierce(request) {
    return await this.#engine.pierce(request)
  }
  async actions(path = '.') {
    const pulse = await this.latest(path)
    const covenantPath = pulse.getCovenantPath()
    const covenantPulse = await this.latest(covenantPath)
    const state = covenantPulse.getState().toJS()
    const { api = {} } = state
    const actions = schemaToFunctions(api)
    const dispatches = {}
    for (const key of Object.keys(actions)) {
      dispatches[key] = (payload) => {
        const action = actions[key](payload)
        return this.dispatch(action, path)
      }
    }
    return dispatches
  }
  async latest(path = '.') {
    const { wd } = this
    const absPath = posix.resolve(wd, path)
    for await (const pulse of this.#engine.subscribe()) {
      debug('latest pulse for', path, pulse.getPulseLink())
      try {
        const latest = await this.#engine.latestByPath(absPath, pulse)
        return latest
      } catch (error) {
        debug('latest error', error.message)
      }
    }
  }
  /**
   * Get whatever pulse is available in local storage for the given path.
   * Will error if no pulse is available.
   * @param {string} path Path to requested pulse relative to rootPulse
   * @param {Pulse} rootPulse Pulse to be treated as root for the path
   * @returns Pulse
   */
  async current(path = '.', rootPulse) {
    const { wd } = this
    const absPath = posix.resolve(wd, path)
    const latest = await this.#engine.latestByPath(absPath, rootPulse)
    return latest
  }
  get pulseResolver() {
    return (pulseLink) => this.#endurance.recover(pulseLink)
  }
  get logger() {
    return this.#engine.logger
  }
  get wd() {
    const { wd = '/' } = this.#engine.selfLatest.getState().toJS()
    return wd
  }
  async hardReset() {
    await this.stop()
    if (isBrowser) {
      const dbs = await window.indexedDB.databases()
      const awaits = []
      for (const db of dbs) {
        debug(`deleting`, db)
        const request = window.indexedDB.deleteDatabase(db.name)
        awaits.push(
          new Promise((resolve, reject) => {
            request.onerror = reject
            request.onsuccess = resolve
          }).then(() => debug(`deleted`, db))
        )
      }
      return await Promise.all(awaits)
    }
    if (isNode) {
      if (typeof this.#repo === 'string') {
        debug('deleting directory:', this.#repo)
        const { default: rimraf } = await import('rimraf')

        // sync to purposefully block the thread from any block making
        return rimraf.sync(this.#repo)
      }
      debug('repo was not a filesystem path - skipping deletion')
    }
  }
  async scrub() {
    /**
    To ensure the integrity of the system, invoke scrub to
    check the hashes of all objects at this location,
    and to verify the availability of all objects.
    Use the option --history to check all blocks in the past
    as well as LATEST.  This will incur significant usage fees
    and bandwidth.
     */
    throw new Error('not implemented')
  }
  async startNetwork() {
    if (this.net) {
      await this.net.start()
    }
  }
  async stop() {
    if (this.net) {
      this.#crypto.stop()
      await this.#endurance.stop() // complete all disk writes
      await this.net.stop()
    }
    await this.#engine.stop() // stop all interpulsing
  }
  async stats() {
    if (!this.net) {
      return {}
    }
    return await this.net.stats()
  }
  subscribe(path = '.') {
    // receive a continuous stream of pulses from a given path
    const { wd } = this
    const absPath = posix.resolve(wd, path)
    debug('subscribing to', absPath)
    const sink = pushable({
      objectMode: true,
      onEnd: () => {
        debug('unsubscribe', absPath)
        this.#subscribers.delete(sink)
      },
    })
    this.#subscribers.add(sink)
    const rootEmitter = this.#engine.subscribe()
    const checker = async (source) => {
      let prior
      for await (const rootPulse of source) {
        debug('checker pulse %s', rootPulse.getPulseLink())
        try {
          const latest = await this.#engine.latestByPath(absPath, rootPulse)
          if (prior?.cid.equals(latest.cid)) {
            continue
          }
          prior = latest
          sink.push(latest)
        } catch (error) {
          debug('error', error.message)
        }
      }
    }
    pipe(rootEmitter, checker)
    return sink
  }
  /**
   * get the latest mtab pulse, then:
   *    get the new multi addresses
   *    get the new address mappings
   *    get the new hardlinks
   */
  async #watchMtab() {
    assert(this.net)
    const subscribed = new Set()
    const multiaddrsSet = new Set()
    let lastMtab
    for await (const mtab of this.subscribe('/.mtab')) {
      if (lastMtab) {
        // TODO determine what to unsubscribe and unmap
      }
      const state = mtab.getState().toJS()
      debug('mtab state', state)
      debug('mtab hash', mtab.cid.toString())
      const peerIds = Object.keys(state)
      for (const peerId of peerIds) {
        const { multiaddrs, chainIds } = state[peerId]
        for (const multiaddr of multiaddrs) {
          if (!multiaddrsSet.has(multiaddr)) {
            multiaddrsSet.add(multiaddr)
            await this.net.addMultiAddress(multiaddr)
          }
        }
        for (const chainId of chainIds) {
          this.net.addAddressPeer(chainId, peerId)
        }
      }

      lastMtab = mtab
      const { hardlinks, channels } = mtab.getNetwork()
      for await (const [, channelId] of hardlinks.entries()) {
        const channel = await channels.getChannel(channelId)
        const { address } = channel
        if (subscribed.has(address.getChainId())) {
          continue
        }
        subscribed.add(address.getChainId())
        debug('subscribing to', address)

        const stream = this.net.subscribePulse(address)
        const updater = async (source) => {
          for await (const pulselink of source) {
            assert(pulselink instanceof PulseLink)
            const latest = await this.#endurance.recover(pulselink)
            const target = mtab.getAddress()
            await this.#engine.updateLatest(target, latest)
          }
        }
        pipe(stream, updater)
      }
    }
  }
  get isCreated() {
    return !this.net || this.net.isCreated
  }
}
const mapShell = (engine) => {
  const actions = {}
  const api = schemaToFunctions(shell.api)
  for (const key of Object.keys(api)) {
    actions[key] = (...args) => {
      const action = api[key](...args)
      return engine.pierce(action)
    }
  }
  const { publish } = actions
  assert(publish)
  actions.publish = (name, covenant = {}, parentPath = '.') => {
    // TODO use the covenant schema to pluck out what is valid
    // or by default strip out any functions
    const { reducer, ...rest } = covenant
    return publish(name, rest, parentPath)
  }
  return actions
}
