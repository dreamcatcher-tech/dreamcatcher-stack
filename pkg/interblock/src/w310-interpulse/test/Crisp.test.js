import { createRamRepo } from '../../w305-libp2p'
import { Interpulse, Crisp, Syncer } from '..'
import { crm } from '../../w301-user-apps'
import { Pulse } from '../../w008-ipld/src'
import Debug from 'debug'
const debug = Debug('tests')

describe('Crisp', function () {
  it('should create a Crisp', async function () {
    const pulse = await Pulse.createCI({ state: { test: true } })
    const crisp = Crisp.createRoot(pulse)
    expect(crisp).toBeInstanceOf(Crisp)
    expect(crisp.state).toEqual({ test: true })
    expect(crisp.hasChild('not a path')).toBe(false)

    // add a child to the pulse
    // dump the engine cache
  })
  it.only('reconciles diffs', async () => {
    const repo = createRamRepo('ram')
    let engine = await Interpulse.createCI({
      overloads: { '/crm': crm.covenant },
      repo,
    })
    await engine.add('app', '/crm')
    await engine.stop()
    engine = await Interpulse.createCI({
      overloads: { '/crm': crm.covenant },
      repo,
    })
    const approot = await engine.current('app')
    Debug.enable('iplog *Crisp *Syncer tests')
    let count = 0
    const resolver = (pl) => ++count && engine.pulseResolver(pl)
    const syncer = Syncer.create(resolver)
    // restart the engine so can do timing
    debug('starting syncer')
    await syncer.update(approot)
    debug('syncer complete', count)

    for await (const crisp of syncer) {
      debug('crisp', crisp)
      const children = [...crisp]
      children.sort((a, b) => a.localeCompare(b))
      debug(children)
      expect(children).toMatchSnapshot()
      break
    }
    await engine.stop()
  })
})
