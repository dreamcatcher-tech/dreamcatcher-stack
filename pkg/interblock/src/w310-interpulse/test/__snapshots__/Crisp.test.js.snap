// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Crisp reconciles diffs 1`] = `
[
  ".",
  "..",
  "about",
  "account",
  "customers",
  "routing",
  "schedule",
  "settings",
]
`;
